freeipa/src/main/java/com/sequenceiq/freeipa/service/freeipa/user/ums/DefaultUmsUsersStateProvider.java

```diff
@@ -0,0 +1,175 @@
+
+        Map<String, FmsGroup> crnToFmsGroup = convertGroupsToFmsGroups(
+                grpcUmsClient.listAllGroups(INTERNAL_ACTOR_CRN, accountId, requestIdOptional));
+        Map<UserManagementProto.WorkloadAdministrationGroup, FmsGroup> wags = convertWagsToFmsGroups(
+                grpcUmsClient.listWorkloadAdministrationGroups(INTERNAL_ACTOR_CRN, accountId, requestIdOptional));
+        List<String> requestedWorkloadUsernames = Streams.concat(
+                users.stream().map(UserManagementProto.User::getWorkloadUsername),
+                machineUsers.stream().map(UserManagementProto.MachineUser::getWorkloadUsername))
+                .collect(Collectors.toList());
+
+        Map<String, UmsUsersState> umsUsersStateMap = new HashMap<>();
+        environmentCrns.forEach(environmentCrn -> {
+            UmsUsersState.Builder umsUsersStateBuilder = new UmsUsersState.Builder()
+                    .setWorkloadAdministrationGroups(wags.values());
+            UsersState.Builder usersStateBuilder = new UsersState.Builder();
+
+            addRequestedWorkloadUsernames(umsUsersStateBuilder, requestedWorkloadUsernames);
+            addGroupsToUsersStateBuilder(usersStateBuilder, crnToFmsGroup.values());
+            Set<String> wagNamesForOtherEnvironments =
+                    addWagsToUsersStateBuilder(usersStateBuilder, wags, environmentCrn);
+
+            ActorHandler actorHandler = ActorHandler.newBuilder()
+                    .withFmsGroupConverter(getFmsGroupConverter())
+                    .withUmsUsersStateBuilder(umsUsersStateBuilder)
+                    .withUsersStateBuilder(usersStateBuilder)
+                    .withCrnToFmsGroup(crnToFmsGroup)
+                    .withWagNamesForOtherEnvironments(wagNamesForOtherEnvironments)
+                    .build();
+            EnvironmentAccessChecker environmentAccessChecker = createEnvironmentAccessChecker(environmentCrn);
+            TriConsumer<String, FmsUser, List<UserManagementProto.CloudIdentity>> actorConsumer =
+                    createActorConsumer(accountId, environmentAccessChecker, actorHandler, requestIdOptional);
+            users.forEach(u -> actorConsumer.accept(u.getCrn(), fmsUserConverter.toFmsUser(u), u.getCloudIdentitiesList()));
+            machineUsers.forEach(mu -> actorConsumer.accept(mu.getCrn(), fmsUserConverter.toFmsUser(mu), mu.getCloudIdentitiesList()));
+
+            addServicePrincipalsCloudIdentities(
+                    umsUsersStateBuilder,
+                    grpcUmsClient.listServicePrincipalCloudIdentities(
+                            INTERNAL_ACTOR_CRN, accountId, environmentCrn, requestIdOptional));
+
+            umsUsersStateBuilder.setUsersState(usersStateBuilder.build());
+            umsUsersStateMap.put(environmentCrn, umsUsersStateBuilder.build());
+        });
+        return umsUsersStateMap;
+    }
+
+    private TriConsumer<String, FmsUser, List<UserManagementProto.CloudIdentity>> createActorConsumer(
```

Comment:
it's a bit strange to depend on something from a logging library which is not log related

Reply:
I'll look for a replacement

Reply:
moved this into a `addActorsToUmsUsersStateBuilder` method to match the Bulk provider. 
I had this TriConsumer so that I could apply the same consumer to both the user and machine users. I switched it around and concat the users and machine users streams so I don't have to save the lambda.